src/main/java/org/wso2/sample/identity/oauth2/grant/ed25519/X509GrantHandler.java

```diff
@@ -87,59 +87,44 @@ public boolean validateGrant(OAuthTokenReqMessageContext oAuthTokenReqMessageCon
-        // retrieve the certificate object
-        byte[] byteArray = Base64.decodeBase64(certificate);
-
-        JSONObject decodedCertificate = decodeCertificate(certificate);
-        String subjectCN = decodedCertificate.getString("username");
-            log.debug("Username is retrieved from subjectCN : " + subjectCN);
+        String[] split = sshCertificate.trim().split("\\s+");
+        byte[] encodedCert = java.util.Base64.getMimeDecoder().decode(split[1].getBytes(UTF_8));
 
-            return subjectCN;
-    }
+        Buffer.PlainBuffer plainBuffer = new Buffer.PlainBuffer(encodedCert);
+        String certificateType;
+        try {
+            certificateType = plainBuffer.readString();
+        } catch (Buffer.BufferException e) {
+            log.error(e);
+            return null;
+        }
 
-    private JSONObject decodeCertificate(String encodedCertificate) {
+        if(SSH_CERT_TYPE_ED25519.equals(certificateType)) {
+            KeyType keyType = KeyType.fromString(SSH_CERT_TYPE_ED25519);
+            PublicKey publicKey;
+            try {
+                publicKey = keyType.readPubKeyFromBuffer(plainBuffer);
 
-        HttpsURLConnection connection = null;
-        String urlParam = "{\"certificate\" : \"" + encodedCertificate + "\"}";
+                if (log.isDebugEnabled()) {
+                    log.debug("PublicKey algorithm " + publicKey.getAlgorithm());
+                }
 
-        try {
-            URL url = new URL("https://localhost:8443/user");
-            connection = (HttpsURLConnection) url.openConnection();
-            connection.setRequestMethod("POST");
-            connection.setRequestProperty("Content-Type", "application/json");
-
-            connection.setRequestProperty("Content-Length", Integer.toString(urlParam.getBytes().length));
-            connection.setRequestProperty("Content-Language", "en-US");
-
-            connection.setUseCaches(false);
-            connection.setDoOutput(true);
-
-            DataOutputStream wr = new DataOutputStream(connection.getOutputStream());
-            wr.writeBytes(urlParam);
-            wr.close();
-
-            InputStream is = connection.getInputStream();
-            BufferedReader rd = new BufferedReader(new InputStreamReader(is));
-            StringBuilder response = new StringBuilder();
-            rd.lines().forEach(response::append);
-            rd.close();
-
-            return new JSONObject(response.toString());
-        } catch (Exception e) {
-            e.printStackTrace();
-        } finally {
-            if (connection != null) {
-                connection.disconnect();
+                if (publicKey instanceof Certificate) {
+                    Certificate certificate = (Certificate) publicKey;
+
+                    List<String> validPrincipals = certificate.getValidPrincipals();
+                    if (validPrincipals != null && !validPrincipals.isEmpty()) {
```

Comment:
We might need to iterate over the principals if any one matches, and then check if that user is not disabled.

Reply:
+1